internal/jimm/jimm.go

```diff
@@ -193,6 +144,92 @@ type OAuthAuthenticator interface {
 	AuthenticateBrowserSession(ctx context.Context, w http.ResponseWriter, req *http.Request) (context.Context, error)
 }
 
+// RoleManager provides a means to manage roles within JIMM.
+type RoleManager interface {
+	// AddRole adds a role to JIMM.
+	AddRole(ctx context.Context, user *openfga.User, roleName string) (*dbmodel.RoleEntry, error)
+	// GetRoleByUUID returns a role based on the provided UUID.
+	GetRoleByUUID(ctx context.Context, user *openfga.User, uuid string) (*dbmodel.RoleEntry, error)
+	// GetRoleByName returns a role based on the provided name.
+	GetRoleByName(ctx context.Context, user *openfga.User, name string) (*dbmodel.RoleEntry, error)
+	// RemoveRole removes the role from JIMM in both the store and authorisation store.
+	RemoveRole(ctx context.Context, user *openfga.User, roleName string) error
+	// RenameRole renames a role in JIMM's DB.
+	RenameRole(ctx context.Context, user *openfga.User, uuid, newName string) error
+	// ListRoles returns a list of roles known to JIMM.
+	// `match` will filter the list fuzzy matching role's name or uuid.
+	ListRoles(ctx context.Context, user *openfga.User, pagination pagination.LimitOffsetPagination, match string) ([]dbmodel.RoleEntry, error)
+	// CountRoles returns the number of roles that exist.
+	CountRoles(ctx context.Context, user *openfga.User) (int, error)
+}
+
+// Option modifies JIMM returned by the constructor.
+type Option func(*JIMM)
+
+// RoleManagerOption replaces the role manager in the returned JIMM with the specified one.
+var RoleManagerOption = func(rm RoleManager) Option {
+	return func(j *JIMM) {
+		j.RoleManager = rm
+	}
+}
+
+// New returns a new instance of JIMM based on the provided parameters and options.
+func New(p Parameters, options ...Option) (*JIMM, error) {
+	j := &JIMM{
+		Parameters: p,
+	}
+
+	if err := j.Database.Migrate(context.Background(), false); err != nil {
+		return nil, errors.E(err)
+	}
+
+	roleManager, err := role.NewRoleManager(j.Database, p.OpenFGAClient)
+	if err != nil {
+		return nil, err
+	}
+	j.RoleManager = roleManager
+	for _, option := range options {
+		option(j)
+	}
+
+	return j, nil
+}
```

Comment:
I think it's correct like this, the JIMM method sets up the business logic, so when we have more "managers", the JIMM struct won't need any of the fields in Params, it will just take the pieces of params like the DB, OpenFGA client, pubSub, etc and pass them to the relevant constructors for the managers. Whoever calls jimm.New() decides what they want to use as a DB, as an auth service, etc (even if you can't mock the DB or auth service today).